### PR TITLE
Remove dynamic setup of `pytest.collect` fake module

### PR DESCRIPTION
**Pull request by blueyed**
Make it a real module instead.

This was initially added in 9b58d6eaca
(https://github.com/pytest-dev/pytest/pull/2315), when removing support
for `pytest_namespace`.

Not sure if it ever worked "properly", but currently
`import pytest.collect` and `from pytest.collect import File` do not
work (but with this patch).

This is only relevant for (backward) compatibility, but the more it
appears to make sense to move it into a lazily loaded, separate file.

### CODE REVIEW

**Comment by blueyed** [comment]
@RonnyPfannschmidt 
What do you mean by "removing the attribute"?
This adds a test for compat that might have even failed before (if being more strict), since now the "attribute" actually behaves like a module.

**Comment by RonnyPfannschmidt** [comment]
But you either have to do eager import or cut a major release

File: src/pytest/__init__.py [hunk 1]
```diff
@@ -95,5 +94,9 @@
 ]
 
 
-_setup_collect_fakemodule()
-del _setup_collect_fakemodule
+def __getattr__(name):
+    if name == "collect":
+        from _pytest.compat import _setup_collect_fakemodule
```
Comment (nicoddemus) [inline src/pytest/__init__.py]: Perhaps follow the previous approach of using a real module (say `_pytest._collect`) and use it here?

```python
def __getattr__(name):
    if name == "collect":
        from _pytest import _collect as collect
        return collect
    raise `AttributeError(name)` 
```

Hmmm unfortunately `__getattr__` is [Python 3.7 only](https://www.python.org/dev/peps/pep-0562/)?

**Comment by nicoddemus** [comment]
> But you either have to do eager import

You mean using a real import on `pytest.__init__`?

**Comment by blueyed** [comment]
The eager import was done in https://github.com/pytest-dev/pytest/pull/6803/commits/8a8b1f67a8b14d892f0ba42f01df5a01bb8923f9.

We can use `__getattr__` for py37+ then, and the old / unconditional code otherwise?

File: src/pytest/__init__.py (hunk 1 above)
Comment (blueyed) [reply src/pytest/__init__.py]: I'd rather keep the original code as-is, given that it is supposed to not stay anyway.

Comment (RonnyPfannschmidt) [reply src/pytest/__init__.py]: As a follow-up or inline, adding to sys. Modules and adding a deprecation warning subclass of the module type should improve the general situation

Comment (blueyed) [reply src/pytest/__init__.py]: Adding to sys.modules would likely break that test: https://github.com/pytest-dev/pytest/pull/6803/files#diff-dc953e3e35889673e6c7ba78247b25d9R49-R50

Comment (blueyed) [reply src/pytest/__init__.py]: The deprecation warning should be done in a follow-up.

**Comment by blueyed** [comment]
Missing coverage is due to `symlink_or_skip` - pending investigation by a Windows user (it seems like all of the CI jobs can create symlinks? - https://github.com/pytest-dev/pytest/pull/6733).

**Comment by nicoddemus** [comment]
> We can use __getattr__ for py37+ then, and the old / unconditional code otherwise?

Good idea, this way we can upgrade to the py37+ only version when the time comes to drop py36.

> Missing coverage is due to symlink_or_skip - pending investigation by a Windows user 

I gave this another go, but no luck yet (commented on that PR).

**Comment by nicoddemus** [comment]
> Missing coverage is due to symlink_or_skip - pending investigation by a Windows user

Btw I don't think that should block this.

**Comment by blueyed** [comment]
> > Missing coverage is due to symlink_or_skip - pending investigation by a Windows user
> 
> Btw I don't think that should block this.

Me neither - but would be good that it works / is covered for non-admin users.

**Pull request closed without merging**